cocotb/_sim_versions.py

```diff
@@ -6,39 +6,108 @@
 Classes to compare simulation versions.
 
 These are for cocotb-internal use only.
-
-.. warning::
-    These classes silently allow comparing versions of different simulators.
 """
 
 from distutils.version import LooseVersion
 
 
-class ActivehdlVersion(LooseVersion):
+class _LooseVersionTypeChecked(LooseVersion):
+    """Do a strict type check so that versions of different simulators cannot be compared.
+
+    Example:
+        >>> IcarusVersion("10.3 (stable)") == VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") != VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") > VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") >= VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") < VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+        >>> IcarusVersion("10.3 (stable)") <= VerilatorVersion("10.3 (stable)")
+        Traceback (most recent call last):
+        ...
+        TypeError: Comparing versions of different simulators doesn't make sense
+    """
+
+    def __eq__(self, other):
+        if type(other) is not type(self):
```

Comment:
I think it may be better if is the type check isn't so strict. This allows comparisons between common subtypes.
```suggestion
        if not isinstance(self, type(other)) and not isinstance(other, type(self)):
```

Reply:
Done.